Sources/PackageCollections/Providers/JSONPackageCollectionProvider.swift

```diff
@@ -0,0 +1,302 @@
+        } else {
+            self.decoder.dateDecodingStrategy = .customISO8601
+        }
+        #endif
+    }
+
+    func get(_ source: PackageCollectionsModel.CollectionSource, callback: @escaping (Result<PackageCollectionsModel.Collection, Error>) -> Void) {
+        guard case .json = source.type else {
+            preconditionFailure("JSONPackageCollectionProvider can only be used for fetching 'json' package collections")
+        }
+
+        if let errors = source.validate() {
+            return callback(.failure(MultipleErrors(errors)))
+        }
+
+        // first do a head request to check content size compared to the maximumSizeInBytes constraint
+        let headOptions = self.makeRequestOptions(validResponseCodes: [200])
+        self.httpClient.head(source.url, options: headOptions) { result in
+            switch result {
+            case .failure(let error):
+                callback(.failure(error))
+            case .success(let response):
+                if let contentLength = response.headers.get("Content-Length").first.flatMap(Int.init),
+                    contentLength >= self.configuration.maximumSizeInBytes {
+                    return callback(.failure(Errors.responseTooLarge(contentLength)))
+                }
+                // next do a get request to get the actual content
+                let getOptions = self.makeRequestOptions(validResponseCodes: [200])
+                self.httpClient.get(source.url, options: getOptions) { result in
+                    switch result {
+                    case .failure(let error):
+                        callback(.failure(error))
+                    case .success(let response):
+                        // check content length again so we can record this as a bad actor
+                        // if not returning head and exceeding size
+                        // TODO: store bad actors to prevent server DoS
+                        guard let contentLength = response.headers.get("Content-Length").first.flatMap(Int.init) else {
+                            return callback(.failure(Errors.invalidResponse("Missing Content-Length header")))
+                        }
+                        guard contentLength < self.configuration.maximumSizeInBytes else {
+                            return callback(.failure(Errors.responseTooLarge(contentLength)))
+                        }
+                        // construct result
+                        callback(makeCollection(response))
+                    }
+                }
+            }
+        }
+
+        func makeCollection(_ response: HTTPClientResponse) -> Result<PackageCollectionsModel.Collection, Error> {
+            let collection: CollectionV1
+            do {
+                // parse json
+                guard let decoded = try response.decodeBody(CollectionV1.self, using: self.decoder) else {
+                    throw Errors.invalidResponse("Invalid body")
```

Comment:
Would there be a way to include the JSON in the error for debugging purposes?

Reply:
Oh, I guess the `decode` itself would throw for syntax errors, and that would include information about the problem.

Reply:
in this case the body is empty